### PR TITLE
Create new view of cell in cell context menu

### LINKED ISSUE #2073: Replace mirrored output by mirrored cells with collapsable input
A single cell can have multiple outputs. Each and any of those outputs (even a single number) can be dragged into the dock panel as a separate widget. The more I have used this, I feel this isn't quite the right model as you end up with mirrored output that entirely lacks context of sibling outputs and input.

I think the user experience will be improved if we:

* Remove the ability to drag individual outputs from a cell.
* Instead, allow the entire cell to be dragged out into the dock panel.
* Also allow inputs to be collapsed for any cell (we are working on that UI).

Then a user can drag a full cell out into the UI and still have the context, but be able to hide/show the input as needed.

### PR DESCRIPTION
**Pull request by jasongrout**
This adds a "Create new view of cell" context menu entry to make a new cell widget tied to the same model.

It exposes a `createCell` command on the notebook object - thoughts on if this should be exposed as a public method? It seems like we took pains to make cell creation private, but it just means we need to duplicate that code when we do want the notebook to create a new cell and hand it to us. But perhaps this is an esoteric enough case that it doesn't warrant the API addition to notebook.

Fixes #2073

### CODE REVIEW

**Comment by jasongrout** [comment]
CC @SylvainCorlay, @ssunkara1, @dmadaka

**Comment by blink1073** [comment]
cc @dmadeka

**Comment by blink1073** [comment]
We had previously considered adding a `.clone()` function to a cell widget that we could also use in [cell tools](https://github.com/jupyterlab/jupyterlab/blob/79ddfae3606602b20259983d50e2b86ca89d7b9f/packages/notebook/src/celltools.ts#L352-L368)

**Comment by blink1073** [comment]
Actually, we should probably have different logic for that that is shared with creating a drag image for a [cell](https://github.com/jupyterlab/jupyterlab/blob/79ddfae3606602b20259983d50e2b86ca89d7b9f/packages/notebook/src/widget.ts#L1564).  Either way, I think `Cell.clone()` makes sense.

**Comment by jasongrout** [comment]
I also thought about doing the clone at the cell level, but the *notebook* is the one that has the factory for creating cell widgets from models, and it may modify the cell widget after it is created (for example, it adds a class to code cells currently).

**Comment by jasongrout** [comment]
> the notebook is the one that has the factory for creating cell widgets from models

Though a counterpoint is that the notebook cell factory may make assumptions about the cell being part of the notebook, rather than in a different tab.

**Comment by blink1073** [comment]
In our case we wouldn't want that class because it is no longer part of a notebook.

Sent from phone

________________________________

**Comment by jasongrout** [comment]
@blink1073 - comments now?

File: packages/cells/src/widget.ts [hunk 1]
```diff
@@ -312,6 +312,17 @@ class Cell extends Widget {
   }
 
   /**
+   * Clone the cell, using the same model.
+   */
+  clone() {
```
Comment (blink1073) [inline packages/cells/src/widget.ts]: Missing a return type.

**Comment by blink1073** [comment]
I like it!

**Comment by blink1073** [comment]
With https://github.com/jupyter-widgets/ipywidgets/pull/1779, we can verify a linked widget:

<image src="https://user-images.githubusercontent.com/2096628/32102436-e8995b06-bae1-11e7-9b1e-e25f57c85550.png" width=400>

**Comment by maartenbreddels** [comment]
This will make many people really happy :) At least me! 👍

File: packages/cells/src/widget.ts (hunk 1 above)
Comment (jasongrout) [reply packages/cells/src/widget.ts]: I couldn't figure out to make the type `this`, so I added explicit types, which meant I needed to add an explicit clone method to RawCell to get the typing right.

**Comment by jasongrout** [comment]
> With jupyter-widgets/ipywidgets#1779, we can verify a linked widget:

Nice! Thanks for that PR.

**Comment by jasongrout** [comment]
One glitch I noticed with this - the CodeMirror refresh() does not seem to be taking effect at the very start.

1. Create a new view of a cell
2. click on the new Cell tab

What I see is the codemirror not auto-sized for the code. As soon as I click inside the codemirror, or move the tab, codemirror auto-fits to the code size (presumably because another refresh() call).

I traced it a bit yesterday - the codemirror refresh is called `onShow`, but perhaps the codemirror is not ready to be refreshed at that point?

**Comment by blink1073** [comment]
When the widget is shown it is not yet properly sized.  This change defers the refresh to allow the CSS rules to be applied:

```diff
diff --git a/packages/codeeditor/src/widget.ts b/packages/codeeditor/src/widget.ts
index 8fca1c930..7743e6b3a 100644
--- a/packages/codeeditor/src/widget.ts
+++ b/packages/codeeditor/src/widget.ts
@@ -76,7 +76,7 @@ class CodeEditorWrapper extends Widget {
   protected onAfterAttach(msg: Message): void {
     super.onAfterAttach(msg);
     if (this.isVisible) {
-      this.editor.refresh();
+      this.update();
     }
   }

@@ -84,7 +84,7 @@ class CodeEditorWrapper extends Widget {
    * A message handler invoked on an `'after-show'` message.
    */
   protected onAfterShow(msg: Message): void {
-    this.editor.refresh();
+    this.update();
   }

   /**
@@ -98,6 +98,10 @@ class CodeEditorWrapper extends Widget {
     }
   }

+  protected onUpdateRequest(msg: Message): void {
+      this.editor.refresh();
+  }
+
   /**
    * Handle a change in model selections.
    */
```

**Comment by jasongrout** [comment]
@blink1073 - that works great! I committed and pushed to this PR.

**Comment by jasongrout** [comment]
I think this is good now.

**Comment by blink1073** [comment]
I submitted a PR against yours.

**Comment by jasongrout** [comment]
Thanks. I think you could have just pushed to my branch as well.

**Comment by blink1073** [comment]
Still one more test to fix, upcoming.

**Comment by jasongrout** [comment]
feel free to just push to my branch.

**Comment by blink1073** [comment]
Yep, should be good to go now.

**Comment by blink1073** [comment]
Thanks!

**Pull request merged**